libraries/AP_Networking/AP_Networking_port.cpp

```diff
@@ -342,22 +347,33 @@ bool AP_Networking::Port::send_receive(void)
 
     if (connected) {
         // handle outgoing packets
-        uint32_t available = writebuffer->available();
-        available = MIN(300U, available);
+        uint32_t available;
+
+        {
+            WITH_SEMAPHORE(sem);
+            available = writebuffer->available();
+            available = MIN(300U, available);
 #if HAL_GCS_ENABLED
-        if (packetise) {
-            available = mavlink_packetise(*writebuffer, available);
-        }
+            if (packetise) {
+                available = mavlink_packetise(*writebuffer, available);
+            }
 #endif
-        if (available > 0) {
-            uint8_t buf[available];
-            auto n = writebuffer->peekbytes(buf, available);
-            if (n > 0) {
-                const auto ret = sock->send(buf, n);
-                if (ret > 0) {
-                    writebuffer->advance(ret);
-                    active = true;
-                }
+            if (available == 0) {
+                return active;
+            }
+        }
+        uint8_t buf[available];
+        uint32_t n;
+        {
+            WITH_SEMAPHORE(sem);
+            n = writebuffer->peekbytes(buf, available);
+        }
+        if (n > 0) {
+            const auto ret = sock->send(buf, n);
```

Comment:
seems like the whole point of this is to hold the semaphore during prep but not during the send?

Reply:
we need a sem when operating on the ring buffers, but not during the socket operations, as socket operations take the global TCP/IP semaphore